### DISCUSSION TITLE
How to train VITS with Automatic Mixed Precision (AMP) training succesfully

### DISCUSSION THREAD

**Discussion by mindmapper15** [espnet/espnet#5293] [General]
I opened this discussion to find out how to train VITS TTS model with mixed precision training without exploding loss.
Any idea that helps to solve this issue will be welcomed!

**Comment by mindmapper15** [comment]
I failed to train VITS TTS in AMP with LJSpeech dataset, even I used the default configuraion in egs2 recipe except doubled batch_bins.
The Discriminator loss goes to nan since 19 epoch.

I attached my train.log and conf.yaml in here.

[train.log](https://github.com/espnet/espnet/files/12011162/train.log)
[config.yaml](https://github.com/espnet/espnet/files/12011165/config.yaml.txt)

I'll try the [original repository](https://github.com/jaywalnut310/vits) made by the paper authors and compare its result with espnet's.

My Training Environment : 
- 4 GPU (NVIDIA Titan RTX)
- torch 1.13.1
- CUDA 11.7

**Comment by mindmapper15** [comment]
AMP training with LJSpeech using the code from [the original repository with fixed loss function](https://github.com/boltzmann-Li/vits) works flawless without any nan loss.

Currently, I trained the model to 26k steps and here is the log from training 
[train.log](https://github.com/espnet/espnet/files/12022585/train.log)

I will compare what is different between the original repo and espnet.

**Comment by sw005320** [comment]
Thanks for your report. 
This is an interesting discussion.
@kan-bayashi, it would be great if you have any comments about this.

**Comment by kan-bayashi** [comment]
Actually, I thought I fixed the issue in https://github.com/espnet/espnet/pull/4356.
But I've never checked the behavior so it might we need to skip the casting somewhere.

**Comment by mindmapper15** [comment]
> Actually, I thought I fixed the issue in #4356. But I've never checked the behavior so it might we need to skip the casting somewhere.

https://github.com/boltzmann-Li/vits/blob/main/train.py#L142-L177
![image](https://github.com/espnet/espnet/assets/40658807/bb93bb00-e2bd-4248-8ebc-6cbcb1234e2a)

Code from the fixed version of the original repository.
The Discriminator is forwarded at outside of autocast context.
I'm not sure this is right, but maybe this is the reason why training with this code didn't explode...?

I'll put back discriminator into autocast context and test again with same dataset.
Will be back with the result.
